File: src/main/java/com/ampro/weebot/Launcher.java
/**
 * @author Jonathan Augustine
 * @copyright Aquatic Mastery Productions
 */

package com.ampro.weebot;

import com.ampro.weebot.commands.ChatbotCommand;
import com.ampro.weebot.commands.Command;
import com.ampro.weebot.commands.HelpCommand;
import com.ampro.weebot.commands.NotePadCommand;
import com.ampro.weebot.commands.developer.DatabaseFileCommand;
import com.ampro.weebot.commands.developer.ListGuildsCommand;
import com.ampro.weebot.commands.developer.ShutdownCommand;
import com.ampro.weebot.commands.developer.WeebotSuggestionCommand;
import com.ampro.weebot.commands.games.SecretePhraseCommand;
import com.ampro.weebot.commands.games.cardgame.CardsAgainstHumanityCommand;
import com.ampro.weebot.commands.management.AutoAdminCommand;
import com.ampro.weebot.commands.management.ManageSettingsCommand;
import com.ampro.weebot.commands.miscellaneous.*;
import com.ampro.weebot.database.Database;
import com.ampro.weebot.database.DatabaseManager;
import com.ampro.weebot.entities.bot.GlobalWeebot;
import com.ampro.weebot.entities.bot.Weebot;
import com.ampro.weebot.jda.JDABuilder;
import com.ampro.weebot.listener.EventDispatcher;
import com.ampro.weebot.util.Logger;
import com.ampro.weebot.util.io.FileManager;
import net.dv8tion.jda.core.AccountType;
import net.dv8tion.jda.core.EmbedBuilder;
import net.dv8tion.jda.core.JDA;
import net.dv8tion.jda.core.JDA.Status;
import net.dv8tion.jda.core.entities.Game;
import net.dv8tion.jda.core.entities.Guild;
import net.dv8tion.jda.core.exceptions.RateLimitedException;

import javax.security.auth.login.LoginException;
import java.awt.*;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static com.ampro.weebot.database.DatabaseManager.DIR_DBS;
import static com.ampro.weebot.util.Logger.LOGS;
import static com.ampro.weebot.util.io.FileManager.*;

/**
 * Runner/Main class of the Weebot network.
 * Builds single JDA connection, instances of all {@link Command Commands}, and
 * {@link Database}.
 *
 * @author Jonathan Augustine
 */
public class Launcher {

	private static final String TOKEN_WBT
			= "NDM3ODUxODk2MjYzMjEzMDU2.DcN_lA.Etf9Q9wuk1YCUnUox0IbIon1dUk";
	private static final String TOKEN_TEST
            = "NDQ0MzIzNzMyMDEwMzAzNDg4.DdaQyQ.ztloAQmeuUffaC-DC9zE-LFwPq4";

	private static JDA JDA_CLIENT;
	private static Thread saveTimer;

	public static GlobalWeebot GLOBAL_WEEBOT;

	private static final ArrayList<Command> COMMANDS =
			new ArrayList<>(Arrays.asList(
					new HelpCommand(), new ShutdownCommand(), new DatabaseFileCommand(),
					new AutoAdminCommand(),new ManageSettingsCommand(),
					new ListGuildsCommand(), new PingCommand(), new SpamCommand(),
					new NotePadCommand(), new SelfDestructMessageCommand(),
					new SecretePhraseCommand(), new WeebotSuggestionCommand(),
					new CardsAgainstHumanityCommand(), new OutHouseCommand(),
                    new ChatbotCommand(), new CalculatorCommand(), new ReminderCommand()
			));

	/** The database */
	private static Database DATABASE;

	/**
	 * Put bot online, setup listeners, and get full list of servers (Guilds)
	 * @param args
	 * @throws LoginException
	 * @throws RateLimitedException
	 * @throws InterruptedException
	 */
   public static void main(final String[] args)
   throws LoginException, InterruptedException {
       System.err.println("[Launcher] Building Directories...");
       if(!Launcher.buildDirs()) {
           System.err.println("[Launcher]\tFAILED!");
           System.exit(-1);
       }
       System.err.println("[Launcher] Initializing Logger...");
       if(!Logger.init()) {
           System.err.println("[Launcher]\tLogger failed to initialize!");
       }
       //Debug
       //RestAction.setPassContext(true); // enable context by default
       //RestAction.DEFAULT_FAILURE = Throwable::printStackTrace;
       Launcher.jdaLogIn();
       //Launcher.jdaDevLogIn();
       Launcher.setUpDatabase();
       Launcher.updateWeebots();
       Launcher.startSaveTimer(.5);
       Launcher.addListeners();

       Logger.derr("[Launcher] Initialization Complete!\n\n");
   }

    /**
     * Build the all file directories used by Weebot.
     * @return {@code false} if any directory is not created (and did not already exist)
     */
    private static boolean buildDirs() {
        if (!DIR_DBS.exists())
            DIR_DBS.mkdirs();
        if (!LOGS.exists())
            LOGS.mkdirs();
        if (!TEMP_OUT.exists())
            TEMP_OUT.mkdirs();
        if (!TEMP_IN.exists())
            TEMP_IN.mkdirs();
        return DIR_HOME.exists() && TEMP_OUT.exists() && TEMP_IN.exists() && DIR_DBS
                .exists() && LOGS.exists();
    }

   /**
    * Initiates {@code Launcher} data and connects to Weebot API.
    * Builds Asynchronously in a separate thread to let main thread work on
    * other setup processes.
    * @throws LoginException
    * @throws InterruptedException
    */
   private static void jdaLogIn()
   throws LoginException, InterruptedException {
       //Connect to API
       Logger.derr("[Launcher] Logging in to Weebot JDA client...");
       JDABuilder builder = new JDABuilder(AccountType.BOT)
                   .setToken(TOKEN_WBT).setGame(Game.playing("@Weebot help"));
       JDA_CLIENT = builder.buildBlocking(Status.CONNECTED);
   }

	/**
	 * Initiates {@code Launcher} data and connects to Weebot TestBuild API.
	 * Builds Asynchronously in a separate thread to let main thread work on
	 * other setup processes.
	 * @throws LoginException
	 * @throws InterruptedException
	 */
	private static void jdaDevLogIn()
    throws LoginException, InterruptedException {
        //Connect to API
        Logger.derr("[Launcher] Logging in to TestBot JDA client...");
        JDABuilder builder = new JDABuilder(AccountType.BOT).setToken(TOKEN_TEST);
        Launcher.JDA_CLIENT = builder.buildBlocking(Status.CONNECTED);
	}

	/**
	 * Adds event listeners to the JDA.
	 */
	private static void addListeners() {
        Logger.derr("[Launcher] Adding Listeners to JDA Client...");
   		Launcher.JDA_CLIENT.addEventListener(new EventDispatcher(), GLOBAL_WEEBOT);
   }

	/**
	 * Attempts to load a database from file. <br>
	 * If a database could not be loaded, a new one is created. <br>
	 * Is called only once during setup.
	 */
	private static void setUpDatabase() {
	    String f = "[Launcher#setUpDatabase]";
        Logger.derr(f + " Setting up Database...");
		Logger.derr(f + "\tLoading database...");
		Database db = DatabaseManager.load();
		Launcher.DATABASE = db == null ? new Database() : db;
		if (db == null) {
            Logger.derr(f + "\t\tUnable to load database, creating new database.");
			Logger.derr(f + "\t\tLoading known Guilds");
			List<Guild> guilds = Launcher.JDA_CLIENT.getGuilds();
			for (Guild g : guilds) {
				Launcher.DATABASE.addBot(new Weebot(g));
			}
			DatabaseManager.save(Launcher.DATABASE);
			Logger.derr(f + "\tDatabase created and saved to file.");
		} else {
			Logger.derr(f + "\tDatabase located.");
            Logger.derr(f + "\tUpdating registered Guilds.");
			Launcher.updateGuilds();
		}
		Logger.derr(f + "\tBacking up database.");
		DatabaseManager.backUp(Launcher.DATABASE);
		Launcher.GLOBAL_WEEBOT = DATABASE.getGlobalWeebot() == null
                        ? new GlobalWeebot() : DATABASE.getGlobalWeebot();
	}

    /**
     * Update the Weebots in the database after downtime.
     * <b>This is only called once on startup</b>
     */
    private static void updateGuilds() {
        List<Guild> guilds = Launcher.JDA_CLIENT.getGuilds();
        guilds.forEach(g -> Launcher.DATABASE.addBot(new Weebot(g)));
    }

	/**
	 * Calls the update method for each Weebot to setup NickNames
	 * changed during downtime and initialize transient variables.
	 */
	private static void updateWeebots() {
        Logger.derr("[Launcher] Updating Weebots...");
	    DATABASE.getWeebots().forEach( (id, bot) -> bot.startup() );
		DATABASE.getGlobalWeebot().startup();
    }

	/**
	 * Starts a thread that saves a database backup each interval.
	 * <br> Listens for a shutdown event to save the the main file
	 * @param min The delay in minuets between saves.
	 */
	private static synchronized void startSaveTimer(double min) {
	   Launcher.saveTimer = new Thread( () -> {
		   int i = 1;
		   long time = Math.round((1000 * 60) * min);
		   try {
			   while (true) {
			       if (Launcher.JDA_CLIENT.getStatus()  == Status.SHUTDOWN)
			           break;
				   if(Launcher.getJda().getStatus() != Status.CONNECTED)
					   continue;
				   DatabaseManager.backUp(Launcher.DATABASE);
				   if(i % 10 == 0) {
                       System.err.println("Database back up: " + i);
                   }
                   i++;
                   Thread.sleep(time);
			   }
		   } catch (InterruptedException e) {
			   e.printStackTrace();
		   }
	   });
	   saveTimer.setName("Save Timer");
	   Launcher.saveTimer.start();
   }

	/** Begin the shutdown sequence. Backup and save database. */
	public static void shutdown() {
	    String f = "[Launcher#shutdown]";
        Logger.derr(f + " Shutdown signal received.");
        Logger.derr(f + "\tClearing registered event listeners...");
        for (Object o : Launcher.JDA_CLIENT.getRegisteredListeners())
		    JDA_CLIENT.removeEventListener(o);
        Logger.derr(f + "\tStopping save timer thread...");
        Launcher.saveTimer.interrupt();
        Logger.derr(f + "\tShutting down Global Weebot Reminder pools...");
		DATABASE.getGlobalWeebot().getReminderPools().forEach(
		        (id, pool) -> pool.shutdown()
        );
        Logger.derr(f + "\tBacking up database...");
		if (DatabaseManager.backUp(Launcher.DATABASE) < 1)
            Logger.derr(f + "\t\tFailed to backup database!");
		else {
            Logger.derr(f + "\tSaving Database...");
            switch (DatabaseManager.save(Launcher.DATABASE)) {
                case -1:
                    Logger.derr(f + "\t\tCould not save backup due to file exception.");
                    break;
                case -2:
                    Logger.derr(f + "\t\tCould not save backup due to corrupt Json.");
                    break;
                default:
                    Logger.derr(f + "\t\tDatabase saved.");
            }
        }

		Logger.derr(f + "\tClearing temp directories...");
		FileManager.clearTempDirs();

        Logger.derr(f + "Successfully shutdown.");

        JDA_CLIENT.shutdown();

	}

	/** @return The acting database. */
	public static Database getDatabase() {
		synchronized (Launcher.DATABASE) {
			return Launcher.DATABASE;
		}
	}

	/**
	 * Check if user ID matches a Developer ID.
	 * @param id long ID to check
	 * @return true if the user ID is a dev.
	 */
	public static boolean checkDevID(long id) {
		return Database.getDevelopers().contains(id);
	}

	/**
	 * Get a guild matching the ID given.
	 * @param id long ID
	 * @return requested Guild <br> null if not found.
	 */
	public static Guild getGuild(long id) {
		Iterable<Guild> it = Launcher.JDA_CLIENT.getGuilds();
		for (Guild g : it)
			if (g.getIdLong() == id)
				return g;

		return null;
	}

	public static ArrayList<Command> getCommands() {
		return Launcher.COMMANDS;
	}

	/**
	 * Get a {@link Command} from the available list.
	 * @param cl The class of the {@link Command}.
	 * @return  The {@link Command} that was requested.
	 *          null if not found.
	 */
	public static Command getCommand(Class<? extends Command> cl) {
		for (Command c : Launcher.COMMANDS) {
			if (c.getClass().equals(cl)) {
				return c;
			}
		}
		return null;
	}

	/** Get the JDA */
	public static JDA getJda() {
		return Launcher.JDA_CLIENT;
	}

	/**@return EmbedBuilder with the standard green, Author set to "Weebot", and footer */
	public static EmbedBuilder getStandardEmbedBuilder() {
		return new EmbedBuilder()
				.setColor(new Color(0x31FF00))
				.setAuthor("Weebot", null, JDA_CLIENT.getSelfUser().getAvatarUrl())
				.setFooter("Run by Weebot", JDA_CLIENT.getSelfUser().getAvatarUrl());
	}

	/**
	 * Makes a standard format EmbedBuilder with standard color and author
	 * , the given title, title URL, and description.
	 * @param title The title of the Embed
	 * @param titleLink The site to link to in the Title
	 * @param description The description that appears under the title
	 * @return A Weebot-standard EmbedBuilder
	 */
	public static EmbedBuilder makeEmbedBuilder(String title,
                                                String titleLink,
                                                String description) {
		return getStandardEmbedBuilder()
				.setTitle(title, titleLink)
				.setDescription(description);

	}

	/**
	 * Makes a standard format EmbedBuilder with standard color and author
	 * , the given title, title URL, and description.
	 * @param title The title of the Embed
	 * @param titleURL The site to link to in the Title
	 * @param description The description that appears under the title
	 * @return A Weebot-standard EmbedBuilder
	 */
	public static EmbedBuilder makeEmbedBuilder(String title, String titleURL,
                                                StringBuilder description) {
		return makeEmbedBuilder(title, titleURL, description.toString());
	}


}

File: src/main/kotlin/com/ampro/weebot/main/Launcher.kt
/*
 * Copyright Aquatic Mastery Productions (c) 2018.
 */

package com.ampro.weebot.main

import com.ampro.weebot.bot.Weebot
import com.ampro.weebot.commands.CMD_HELP
import com.ampro.weebot.commands.commands
import com.ampro.weebot.commands.utilitycommands.remThreadPool
import com.ampro.weebot.database.*
import com.ampro.weebot.database.constants.*
import com.ampro.weebot.extensions.*
import com.ampro.weebot.listeners.EventDispatcher
import com.ampro.weebot.util.*
import com.ampro.weebot.util.Emoji.*
import com.jagrosh.jdautilities.command.CommandClient
import com.jagrosh.jdautilities.command.CommandClientBuilder
import com.jagrosh.jdautilities.commons.waiter.EventWaiter
import kotlinx.coroutines.*
import net.dv8tion.jda.bot.sharding.ShardManager
import net.dv8tion.jda.core.JDA
import net.dv8tion.jda.core.entities.Game.playing
import net.dv8tion.jda.core.entities.SelfUser
import net.dv8tion.jda.core.entities.User
import java.io.File
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit.MILLISECONDS
import java.util.logging.Level.CONFIG
import javax.security.auth.login.LoginException
import kotlin.system.measureTimeMillis

lateinit var SAVE_JOB: Job
/** How ofter to asve to file in Seconds */
const val SAVE_INTER = 30

val CACHED_POOL =  Executors.newCachedThreadPool().asCoroutineDispatcher()

/** Main Logger */
lateinit var MLOG: FileLogger

//JDA connection shard Client
lateinit var JDA_SHARD_MNGR: ShardManager
var ON = true
lateinit var CMD_CLIENT: CommandClient
val WAITER: EventWaiter = EventWaiter()
/** The bot's selfuser from */
lateinit var SELF: SelfUser


/**
 * Put bot online, setup listeners, and get full list of servers (Guilds)
 *
 *
 * @param args_
 * @throws LoginException
 * @throws RateLimitedException/
 * @throws InterruptedException
 */
fun main(args_: Array<String>) {
    runBlocking {
        slog("Launching...")
        slog("\tBuilding Directories...")
        if (!buildDirs()) {
            elog("\tFAILED: Build Dir | shutting down...")
            System.exit(-1)
        }
        slog("\t...DONE")
        slog("Initializing Main Logger")
        MLOG = FileLogger("Launcher $NOW_STR_FILE")
        slog("...DONE\n\n")

        /** Setup for random methods and stuff that is needed before launch */
        val genSetup = listOf(launch { setupWebFuel() })

        //Debug
        //RestAction.setPassContext(true) // enable context by default
        //RestAction.DEFAULT_FAILURE = Consumer(Throwable::printStackTrace)


        //LOGIN & LISTENERS
        val reg_wbot = Regex("(?i)(w|weebot|full)")
        val weebot = (args_.isEmpty() || args_[0].matches(reg_wbot))
                || (FILE_CONFIG.exists() && FILE_CONFIG.readLines()[0].matches(reg_wbot))

        val alt = if (weebot) "\\" else "t\\"


        //COMMAND CLIENT
        CMD_CLIENT = CommandClientBuilder().setOwnerId(DEV_IDS[0].toString())
            .setCoOwnerIds(DEV_IDS[1].toString())
            .setGuildSettingsManager { getWeebotOrNew(it.idLong).settings }
            .setAlternativePrefix(alt)
            //.setGame(listening("@Weebot help"))
            .setGame(playing("Weebot 2.0 Kotlin!"))
            .addCommands(commands)
            .setEmojis(heavy_check_mark.unicode, Warning.unicode, X_Red.unicode)
            .setServerInvite(LINK_INVITEBOT)
            .setHelpConsumer { event ->
                //If the only argument is the command invoke
                val args = event.splitArgs()
                if (args.isEmpty()) {
                    CMD_HELP.execute(event)
                } else {
                    commands.forEach { cmd ->
                        if (cmd.isCommandFor(args[0]) && (!cmd.isHidden || event.isOwner)) {
                            if (cmd.getHelpBiConsumer() != null) {
                                cmd.getHelpBiConsumer().accept(event, cmd)
                                return@setHelpConsumer
                            } else if (!cmd.help.isNullOrBlank()) {
                                event.reply("*${cmd.help}*")
                                return@setHelpConsumer
                            } else {
                                event.reply("*Help is currently unavailable for this command." +
                                        " You can use ``@Weebot sugg`` to send feedback to the" +
                                        " Developers and remind them they have a job to do!* $Rage")
                                return@setHelpConsumer
                            }
                        }
                    }
                }
            }
            .setDiscordBotsKey(BOTSONDISCORD_KEY)
            //.setDiscordBotListKey(BOTLIST_KEY)
            .build()

        //LOGIN & LISTENERS
        JDA_SHARD_MNGR = if (weebot) {
            jdaShardLogIn().addEventListeners(CMD_CLIENT).build()
        } else {
            jdaDevShardLogIn().addEventListeners(CMD_CLIENT).build()
        }

        //WAIT FOR SHARD CONNECT
        MLOG.slog("Shard connected! ${measureTimeMillis {
            while (JDA_SHARD_MNGR.shards[0].status != JDA.Status.CONNECTED) {
                MLOG.slog("Waiting for shard to connect...")
                Thread.sleep(500)
            }
        } / 1_000} seconds")

        //DATABASE
        setUpDatabase()
        //Stats
        setUpStatistics()

        JDA_SHARD_MNGR.addEventListener(EventDispatcher(), WAITER)

        genSetup.joinAll() //Ensure all gensetup is finished

        //SET SELF AND AVATAR URL
        SELF = JDA_SHARD_MNGR.shards[0].selfUser
        weebotAvatar = SELF.avatarUrl

        startupWeebots()

        MLOG.slog("Starting Save Job...")
        SAVE_JOB = saveTimer()

        MLOG.slog("Launch Complete!\n\n")

        JDA_SHARD_MNGR.getTextChannelById(BOT_DEV_CHAT).sendMessage("ONLINE!")
            .queueAfter(850, MILLISECONDS)
    }
}

/**
 * Attempts to load Statistics data from file. Sets [STAT] to the loaded data
 * or makes a new instance
 */
fun setUpStatistics() {
    MLOG.slog("Setting up Statistics...")
    MLOG.slog("\tLoading Statistics...")
    val stat: Statistics? = loadJson<Statistics>(STAT_SAVE)
    if (stat == null) {
        MLOG.slog("\t\tUnable to load Statistics, creating new instance.")
        STAT = Statistics()
        if (STAT.saveJson(STAT_SAVE) == -1) {
            MLOG.slog("\tFAILED")
            return
        }
        MLOG.slog("\tStatistics instance created and saved to file.")
    } else {
        MLOG.slog("\tStatistics located.")
        STAT = stat
    }
    MLOG.slog("\tBacking up Statistics.")
    STAT.saveJson(STAT_BK)
    MLOG.slog("\t...DONE")
    MLOG.slog("...DONE")
}

/**
 * Attempts to loadDao a database from file.
 * If a database could not be loaded, a new one is created.
 * Is called only once during setup.
 */
private fun setUpDatabase() {
    MLOG.slog("Setting up Database...")
    MLOG.slog("\tLoading database...")
    var tdao = loadDao()
    if (tdao == null) {
        MLOG.slog("\t\tUnable to loadDao database, creating new database.")
        tdao = Dao()
        MLOG.slog("\t\tLoading known Guilds")
        JDA_SHARD_MNGR.guilds.forEach { tdao.addBot(Weebot(it)) }
        tdao.save()
        MLOG.slog("\tDatabase created and saved to file.")
        DAO = tdao
    } else {
        MLOG.slog("\tDatabase located. Updating registered Guilds.")
        DAO = tdao
        //Update the Weebots in the database after downtime.
        JDA_SHARD_MNGR.guilds.forEach { DAO.addBot(Weebot(it)) }
    }
    DAO.updatePremiumUsers()
    MLOG.slog("\tBacking up database.")
    DAO.backUp()
    MLOG.slog("\t...DONE")
    MLOG.slog("...DONE")
}

/**
 * Calls the update method for each Weebot to setup NickNames
 * changed during downtime and initialize transient variables.
 */
private fun startupWeebots() {
    MLOG.slog("Starting Weebots...")
    DAO.GLOBAL_WEEBOT.startUp()
    DAO.WEEBOTS.values.forEach { bot -> bot.startUp() }
}

/**
 * Starts a [Job] that saves a database backup each interval.
 * Listens for a shutdown event to save the the main file
 *
 * @param min The delay in minuets between saves.
 */
@Synchronized
private fun saveTimer() = GlobalScope.launch {
    var i = 1
    try {
        while (ON) {
            DAO.backUp()
            STAT.saveJson(STAT_SAVE)
            if (i % 50 == 0) {
                MLOG.slog("Database back up: $i")
            }
            i++
            delay(SAVE_INTER * 1_000L)
        }
    } catch (e: InterruptedException) {
        e.printStackTrace()
        sendSMS(PHONE_JONO, "WEEBOT: Save Job Failed")
    }
}

/** Begin the shutdown sequence. Backup and save database.  */
fun shutdown(user: User? = null) {
    if (user != null)
        MLOG.elog("Shutdown signal received from ${user.name} (${user.id}).")

    MLOG.elog("\tShutting down Global Weebot Reminder pools...")
    //DAO.GLOBAL_WEEBOT.reminderPools.forEach { _, pool -> pool.shutdown() }
    MLOG.elog("\tBacking up database...")
    if (DAO.backUp() < 1)
        MLOG.elog("\t\tFailed to backup database!")
    else {
        MLOG.elog("\tSaving Database...")
        when (DAO.save()) {
            -1   -> MLOG.elog("\t\tCould not save backup due to file exception.")
            -2   -> MLOG.elog("\t\tCould not save backup due to corrupt Json.")
            else -> MLOG.elog("\t\tDatabase saved.")
        }
    }

    MLOG.elog("\tClearing temp directories...")
    clearTempDirs()

    JDA_SHARD_MNGR.shutdown()
    JDA_SHARD_MNGR.statuses.forEach { _, status ->
        while (status != JDA.Status.SHUTDOWN) {}
    }

    remThreadPool.close()
    CACHED_POOL.close()

    MLOG.elog("Safely shutdown.")
    System.exit(0)
}

const val GENERIC_ERR_MESG = "*Sorry, I tripped over my shoelaces. Please try that " +
        "again later*"
